### PR TITLE
Increase mdstore Tests

### LINKED ISSUE #4: coda_mdstore Tests
Add tests for `coda_mdstore`.

### PR DESCRIPTION
**Pull request by damonkelley**
Will partially resolve #4 by adding tests for views and urls.

### CODE REVIEW

File: coda/coda_mdstore/tests/test_views.py [hunk 1]
```diff
@@ -0,0 +1,784 @@
...
+class TestAllBagsView:
+
+    @pytest.fixture(autouse=True)
+    def mock_pe(self, monkeypatch):
+        """Patches the paginate_entries with a Mock object."""
+        mock_pe = mock.Mock()
+        monkeypatch.setattr(
+            'coda_mdstore.views.paginate_entries', mock_pe)
+
+        return mock_pe
```
Comment (damonkelley) [inline coda/coda_mdstore/tests/test_views.py]: Attach this to `self` to conform to how the other fixtures are setup.

File: coda/coda_mdstore/tests/test_views.py [hunk 2]
```diff
@@ -0,0 +1,780 @@
...
+
+        # URL with the identifier as a url paramater. We will feed the url to
+        # the request factory because the view will use the path to determine
+        # the id field.
+        url = reverse(
+            'coda_mdstore.views.externalIdentifierSearch', args=[ext_id.value])
+        request = rf.get(url)
+        response1 = views.externalIdentifierSearch(request, ext_id.value)
+
+        # URL with the ark id as a query paramater.
```
Comment (ldko) [inline coda/coda_mdstore/tests/test_views.py]: Not looking for typos really at the moment, but: paramater --> parameter. Also, do you want to capitalize instance of "url" throughout?

Comment (damonkelley) [inline coda/coda_mdstore/tests/test_views.py]: >Not looking for typos really at the moment, but: paramater --> parameter.

haha

>do you want to capitalize instance of "url" throughout?

That is probably a good idea.

File: coda/coda_mdstore/tests/test_views.py [hunk 3]
```diff
@@ -0,0 +1,780 @@
...
+        )
+
+        # Just like in the previous test, feed the real url to the request
+        # factory because the request path is used in the response content.
+        url = reverse(
+            'coda_mdstore.views.externalIdentifierSearch', args=[ext_id.value])
+        request = rf.get(url)
+        response1 = views.externalIdentifierSearch(request, ext_id.value)
+
+        # URL with the ark id as a query paramater.
```
Comment (ldko) [inline coda/coda_mdstore/tests/test_views.py]: "paramater" again

File: coda/coda_mdstore/tests/test_views.py [hunk 4]
```diff
@@ -0,0 +1,780 @@
...
+        assert content[0]['name'] == bag.name
+        assert content[0]['oxum'] == '{0}.{1}'.format(bag.size, bag.files)
+
+    def test_with_with_multiple_identifiers(self, rf):
+        bag = FullBagFactory.create()
+        ext_id = ExternalIdentifierFactory.create(
+            belong_to_bag=bag,
+            value='ark:/67531/metadc000001'
+        )
+        ExternalIdentifierFactory.create(
```
Comment (ldko) [inline coda/coda_mdstore/tests/test_views.py]: Is this supposed to be here?

Comment (damonkelley) [inline coda/coda_mdstore/tests/test_views.py]: yeah, this tests the outcome if there are two identifiers with the same `value`. Only one of the instances need to be kept track of in order though.

**Comment by damonkelley** [comment]
ping @somexpert @ldko 

Made few modifications. This should be ready for review now

File: coda/coda_mdstore/factories.py [hunk 5]
```diff
@@ -0,0 +1,80 @@
+"""
+Coda MDStore Model factories for test fixtures.
+"""
+import factory
+
+from datetime import datetime
```
Comment (somexpert) [inline coda/coda_mdstore/factories.py]: I believe that this and the above import should be switched, per PEP-8.

File: coda/coda_mdstore/factories.py [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+    value = factory.Sequence(lambda n: 'value {0}'.format(n))
+
+
+class FullBagFactory(BagFactory):
+    """
+    Factory to create a Bag object with two related Bag_Info objects, and
+    2 related External_Identifier objects.
+    """
+    info1 = factory.RelatedFactory(
```
Comment (somexpert) [inline coda/coda_mdstore/factories.py]: Need a space after class docstring

File: coda/coda_mdstore/tests/test_utils.py [hunk 7]
```diff
@@ -0,0 +1,14 @@
+from django.core.paginator import Paginator
+import pytest
```
Comment (somexpert) [inline coda/coda_mdstore/tests/test_utils.py]: Should these two imports be reversed and separated with a blank line?

File: coda/coda_mdstore/tests/test_views.py [hunk 8]
```diff
@@ -0,0 +1,890 @@
...
+        response = client.get(reverse('coda_mdstore.views.index'))
+        assert key in response.context[-1]
+
+    def test_renders_correct_template(self, client):
+        response = client.get(reverse('coda_mdstore.views.index'))
+        assert response.template[0].name == 'mdstore/index.html'
+
+    def test_totals_files__sum_key_is_none(self, client):
+        """
+        Check that the `files__sum` key is set 0 if the db returns
```
Comment (somexpert) [inline coda/coda_mdstore/tests/test_views.py]: ...set to 0....

File: coda/coda_mdstore/tests/test_views.py [hunk 9]
```diff
@@ -0,0 +1,890 @@
...
+
+
+class TestStatsView:
+    """
+    Tests for coda_mdstore.views.stats.
+    """
+
+    def test_no_bags_available(self, client):
+        """
+        Test the context values when there are no bags are in the database.
```
Comment (somexpert) [inline coda/coda_mdstore/tests/test_views.py]: should probably remove either 'there are' or the second 'are'.

File: coda/coda_mdstore/tests/test_views.py [hunk 10]
```diff
@@ -0,0 +1,890 @@
...
+        assert response.status_code == 200
+        assert response.context[-1]['totals'] == 0
+        assert response.context[-1]['monthly_bags'] == []
+        assert response.context[-1]['monthly_files'] == []
+        assert response.context[-1]['monthly_running_bag_total'] == []
+        assert response.context[-1]['monthly_running_file_total'] == []
+
+    def test_with_bags(self, client):
+        """
+        Test the context values when there are bags are in the database.
```
Comment (somexpert) [inline coda/coda_mdstore/tests/test_views.py]: same as last comment, should remove either 'there are' or second 'are'

File: coda/coda_mdstore/tests/test_views.py [hunk 11]
```diff
@@ -0,0 +1,890 @@
...
+
+
+class TestExternalIdentiferSearchJSON:
+    """
+    Tests for coda_mdstore.views.externalIdentifierSearchJSON.
+    """
+
+    @pytest.mark.xfail(reason='Exception raised without the `ark` query'
+                              ' parameter.')
+    def test_with_without_ark_parameter(self, rf):
```
Comment (somexpert) [inline coda/coda_mdstore/tests/test_views.py]: Is this test name correct?

**Comment by somexpert** [comment]
I noticed several TODO's, `assert 0`, and expected failures. What's the plan for those?

File: coda/coda_mdstore/tests/test_urls.py [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+    assert resolve('/APP/node/coda-123/').func == views.app_node
+
+
+def test_showNodeStatus():
+    assert resolve('/node/').func == views.showNodeStatus
+
+
+def test_showNodeStatus_with_identifier():
+    assert resolve('/node/coda-123/').func == views.showNodeStatus
+
```
Comment (ldko) [inline coda/coda_mdstore/tests/test_urls.py]: Curious about going from one line to two lines per test for testing same thing.

**Comment by ldko** [comment]
Should there be url tests for:
/resourceindex.xml
/resourcelist-001.xml
/changelist.xml
/capabilitylist.xml
?

File: coda/coda_mdstore/tests/test_views.py [hunk 13]
```diff
@@ -0,0 +1,890 @@
...
+        request = rf.get('/', {'ark': ext_id.value})
+        response = views.externalIdentifierSearchJSON(request)
+        content = json.loads(response.content)
+
+        assert len(content) == 1
+        assert 'bagging_date' in content[0]
+        assert content[0]['name'] == bag.name
+        assert content[0]['oxum'] == '{0}.{1}'.format(bag.size, bag.files)
+
+    def test_with_with_multiple_identifiers(self, rf):
```
Comment (ldko) [inline coda/coda_mdstore/tests/test_views.py]: with_with in name

File: coda/coda_mdstore/tests/test_views.py [hunk 14]
```diff
@@ -0,0 +1,890 @@
...
+        FullBagFactory.create_batch(10)
+
+        request = rf.get('/', HTTP_HOST='example.com')
+        response = views.app_bag(request)
+
+        tree = objectify.fromstring(response.content)
+        assert len(tree.entry) == 10
+
+    @pytest.mark.xfail(reason="Refactor required.")
+    def test_get_request_with_identifier(self):
```
Comment (ldko) [inline coda/coda_mdstore/tests/test_views.py]: I know there is some wonky code in app_bag for a GET request with identifier, but can you either put a test in for a general GET with identifier request in addition to a test marked to fail for parts that will fail, of if you cannot test any path through this, please put a more specific reason.

File: coda/coda_mdstore/tests/test_views.py [hunk 15]
```diff
@@ -0,0 +1,890 @@
...
+
+        request = rf.post('/', HTTP_HOST='example.com')
+        response = views.app_bag(request)
+
+        assert response.status_code == 201
+        assert response['Content-Type'] == 'application/atom+xml'
+        assert response['Location'] == (
+            'http://example.com/APP/bag/{0}/'.format(bag.name)
+        )
+
```
Comment (ldko) [inline coda/coda_mdstore/tests/test_views.py]: We should have a PUT test in here.

File: coda/coda_mdstore/tests/test_views.py [hunk 16]
```diff
@@ -0,0 +1,890 @@
...
+    def test_delete_request_removes_member_ext_identifier_objects(self, rf):
+        bag = FullBagFactory.create()
+        request = rf.delete('/', HTTP_HOST='example.com')
+
+        assert models.External_Identifier.objects.exists() is True
+
+        response = views.app_bag(request, bag.name)
+        assert response.status_code == 200
+        assert models.External_Identifier.objects.exists() is False
```
Comment (ldko) [inline coda/coda_mdstore/tests/test_views.py]: Add test for invalid request method.

**Comment by ldko** [comment]
I didn't see tests for prepare_graph_date_range, calc_total_by_month, paginate_entries, CustomFeed, AtomSiteNewsFeed, bagFullTextSearch, percent. Probably they don't have tests for various reasons at the moment, but listing them here in case you just forgot on any of them.

File: coda/coda_mdstore/factories.py (hunk 6 above)
Comment (damonkelley) [inline coda/coda_mdstore/factories.py]: i disagree :scream_cat: 

I usually only add the blank line below the docstring if a method follows. For class variables, I like them to be directly under the docstring. PEP8 is vague about this.

**Comment by damonkelley** [comment]
>I noticed several TODO's, assert 0, and expected failures. What's the plan for those?

All of the TODOs and `assert 0` should be marked with `@pytest.mark.xfail`, and include a reason for why there is currently not a test. 

I used `assert 0` so that the test will fail, and the runner will reflect that in the output, but the exit status will still be 0 (unless there is a real failure)

**Comment by somexpert** [comment]
Oh yeah, I wasn't saying that they shouldn't be there or anything. Simply curious if the todo's are for a later refactoring, or if they're going to be changed within this or subsequent PRs.

**Comment by damonkelley** [comment]
>I didn't see tests for prepare_graph_date_range, calc_total_by_month, paginate_entries, CustomFeed, AtomSiteNewsFeed, bagFullTextSearch, percent. Probably they don't have tests for various reasons at the moment, but listing them here in case you just forgot on any of them.

That is correct. I am planning to put the feed tests (if there are any) in a `test_feeds.py`, and put the tests for the other functions in `test_utils.py`, because those functions do not actually return an HTTP Response.

File: coda/coda_mdstore/tests/test_urls.py (hunk 12 above)
Comment (damonkelley) [inline coda/coda_mdstore/tests/test_urls.py]: @ldko, can you elaborate?

File: coda/coda_mdstore/tests/test_views.py (hunk 11 above)
Comment (damonkelley) [inline coda/coda_mdstore/tests/test_views.py]: no way

File: coda/coda_mdstore/tests/test_views.py (hunk 16 above)
Comment (damonkelley) [inline coda/coda_mdstore/tests/test_views.py]: So `PATCH` or `HEAD`?

File: coda/coda_mdstore/tests/test_urls.py (hunk 12 above)
Comment (ldko) [inline coda/coda_mdstore/tests/test_urls.py]: Just noting the move from having the test body all on one line:
  ```
  assert resolve('/node/coda-123/').func == views.showNodeStatus
```
to splitting it on to two lines.
```
   url = resolve('/extidentifier/test_value/')
   assert url.func == views.externalIdentifierSearch
```
Inconsistent, but nothing major.

Comment (damonkelley) [inline coda/coda_mdstore/tests/test_urls.py]: ah, yes that was done to keep the lines below the line limit.

File: coda/coda_mdstore/tests/test_views.py (hunk 16 above)
Comment (ldko) [inline coda/coda_mdstore/tests/test_views.py]: If request.method is not 'GET', 'POST', 'PUT', or 'DELETE' the code should return a HttpResponseBadRequest. Doesn't really matter if it is 'PATCH', 'HEAD', or whatever else really.

Comment (damonkelley) [inline coda/coda_mdstore/tests/test_views.py]: I turned `test_delete_request_returns_bad_request` into a general test that will check that a HttpResponseBadRequest is returned.

File: coda/coda_mdstore/tests/test_utils.py (hunk 7 above)
Comment (damonkelley) [inline coda/coda_mdstore/tests/test_utils.py]: @somexpert hashed this out in slack. leaving as is.

File: coda/coda_mdstore/tests/test_views.py (hunk 15 above)
Comment (damonkelley) [inline coda/coda_mdstore/tests/test_views.py]: done in 4502546

**Comment by damonkelley** [comment]
>Should there be url tests for:
/resourceindex.xml
/resourcelist-001.xml
/changelist.xml
/capabilitylist.xml
?


Added in c00545b

**Comment by damonkelley** [comment]
@somexpert, @ldko 

Anything else that needs to be addressed?

File: coda/coda_mdstore/tests/test_urls.py [hunk 17]
```diff
@@ -0,0 +1,107 @@
...
+def test_feed():
+    assert resolve('/feed/').func.__class__ == views.AtomSiteNewsFeed
+
+
+def test_resourceindex():
+    assert resolve('/resourceindex.xml').func == sitemaps.views.index
+
+
+def test_resourcelist_secion():
```
Comment (somexpert) [inline coda/coda_mdstore/tests/test_urls.py]: secion -> sec**t**ion

File: coda/coda_mdstore/tests/test_views.py [hunk 18]
```diff
@@ -0,0 +1,935 @@
...
+
+        assert response.status_code == 200
+        assert response['Content-Type'] == 'application/atom+xml'
+
+        tree = objectify.fromstring(response.content)
+        bag_xml = tree.content[self.CODA_XML]
+        assert bag_xml.name == bag.name
+
+    @pytest.mark.xfail(reason='WSGI request objects do not have a ._req '
+                              'attribtue.')
```
Comment (somexpert) [inline coda/coda_mdstore/tests/test_views.py]: attribtue -> attribute

File: coda/coda_mdstore/tests/test_views.py [hunk 19]
```diff
@@ -0,0 +1,935 @@
...
+    def test_get_request_with_identifier_with_erc_support(self, rf):
+        bag = FullBagFactory.create()
+        request = rf.get('/??', HTTP_HOST='example.com')
+        response = views.app_bag(request, bag.name)
+        assert response['Content-Type'] == 'text/plain'
+
+    @pytest.mark.xfail(reason='WSGI request objects do not have a ._req '
+                              'attribtue.')
```
Comment (somexpert) [inline coda/coda_mdstore/tests/test_views.py]: same here

**Comment by somexpert** [comment]
:thumbsup:

**Comment by ldko** [comment]
:+1:  Thanks!

**Pull request merged**